pysmt/optimization/goal.py

```diff
@@ -0,0 +1,51 @@
+from pysmt.shortcuts import Times, Int
+
+"""
+Abstract class for goals
+"""
+class Goal:
+    pass
+
+"""
+Maximization goal common to all solvers.
+Attention: some solvers may not support this goal
+"""
+class MaximizationGoal(Goal):
+    def __init__(self, formula):
+        self.formula = formula
+
+    def term(self):
+        return self.formula
+
+    def minimize_term(self):
+        return Times(Int(-1), self.formula)
```

Comment:
This needs refinement.
- It's OK if formula is an integer-type
- It would be better to use `Real(-1)` if formula is a real-type 
- If formula is a bit-vector-type you need to use `BVSub(BVZero, formula)`
- you need to accept an environment at the constructor in case a user wants to specify it, and by default use the formula_manager in the default environment (have a look at any solver for an example of this)

Reply:
@PatrickTrentin88 I am not 100%sure about bit vectors: the idea here would be to transform a maximization into a minimization to simplify the implementation of solver-dependent code.

Two questions:
1. Does this transformation impose any performance penalty / do you see limitations in this?
2. Is the transformation above correct for signed and unsigned bit-vectors?

Reply:
@mikand  Such transformation does not appear to be type-correct in the case of a Floating-Point or Bit-Vector goal, since it is mixing LA (`Times`, `Int`) with a `formula` of an arbitrary type. 

I don't see the need to rewrite maximization as minimization when both `z3` and `OptiMathSAT` have native support maximization.

Even if one wanted to implement their own optimization procedures on top of a regular SMT solver, supporting both requires changing only the constraint learned at each iteration. 

What am I not seeing?

Reply:
You are missing nothing :)

I was thinking of starting by forcing the use of minimization consistently to preserve the simplicity of the optimizers "from outside" (the ones that incrementally search for better solutions) that currently support only minimization, but maybe it is better di directly support minimization **and** maximization without converting one into the other.

Reply:
@mikand @arrighi79

Unless in `pysmt` there is a universal `LT` constraint, which translates to the appropriate SMT-LIB operator given the type of its arguments, the simplicity is already lost when one handles objectives of different types. 

How about:

```
    def make_better_than(goal, value):
        """returns (goal < value) when minimizing and
            (value < goal) when maximizing"""
        ...
```

The method would also be responsible for checking for the type of `goal.formula` and use the SMT-LIB operator of the right type.

Then the incremental optimization search becomes completely agnostic about the internal details of the `Goal` instance(s) it is handling.